modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataOutputDelegate.java

```diff
@@ -0,0 +1,127 @@
+
+    OutputMetadataDescriptor descriptor =
+        outputDescriptor().withReturnType(outputDescriptor).withAttributesType(attributesDescriptor).build();
+
+    if (!outputMetadataResult.isSuccess() || !attributesMetadataResult.isSuccess()) {
+      return mergeFailures(descriptor, outputMetadataResult, attributesMetadataResult);
+    }
+
+    return success(descriptor);
+  }
+
+  /**
+   * Given a {@link MetadataKey} of a type and a {@link MetadataContext}, resolves the {@link MetadataType} of the Components's
+   * output using the {@link OutputTypeResolver} associated to the current component.
+   *
+   * @param context {@link MetadataContext} of the Metadata resolution
+   * @param key     {@link MetadataKey} of the type which's structure has to be resolved
+   * @return a {@link MetadataResult} with the {@link MetadataType} of the component's output
+   */
+  private MetadataResult<MetadataType> getOutputMetadata(final MetadataContext context, final Object key) {
+    OutputModel output = component.getOutput();
+    if (isVoid(output.getType()) || !output.hasDynamicType()) {
+      return success(subTypesUnion(output.getType(), typesMapping));
+    }
+
+    return resolveMetadataType(false, subTypesUnion(output.getType(), typesMapping),
+                               () -> resolverFactory.getOutputResolver().getOutputType(context, key), "Output");
+  }
+
+  /**
+   * Given a {@link MetadataKey} of a type and a {@link MetadataContext}, resolves the {@link MetadataType} of the Components's
+   * output {@link Message#getAttributes()} using the {@link OutputTypeResolver} associated to the current component.
+   *
+   * @param context {@link MetadataContext} of the Metadata resolution
+   * @param key     {@link MetadataKey} of the type which's structure has to be resolved
+   * @return a {@link MetadataResult} with the {@link MetadataType} of the components output {@link Message#getAttributes()}
+   */
+  private MetadataResult<MetadataType> getOutputAttributesMetadata(final MetadataContext context, Object key) {
+    OutputModel attributes = component.getOutputAttributes();
+    if (isVoid(attributes.getType()) || !attributes.hasDynamicType()) {
+      return success(subTypesUnion(attributes.getType(), typesMapping));
+    }
+
+    return resolveMetadataType(false, subTypesUnion(attributes.getType(), typesMapping),
+                               () -> resolverFactory.getOutputAttributesResolver().getAttributesMetadata(context, key),
+                               "OutputAttributes");
+  }
+
+  private MetadataResult<TypeMetadataDescriptor> toTypeMetadataDescriptorResult(MetadataType type,
+                                                                                MetadataResult<MetadataType> result) {
+    MetadataType resultingType = result.get() == null ? type : result.get();
+
+    TypeMetadataDescriptor descriptor = typeDescriptor()
+        .withType(resultingType)
+        .dynamic(!type.equals(resultingType))
```

Comment:
I get what you meant about the `equals` returning false on the slightest change, but that is actually what I intended.
If the `type` is not exactly the same type returned by `component.getOutputAttributes().getType()` then it means that some kind of dynamic resolver was involved, and if that happens, then we are in a dynamic case.
(it will even be the same type instance, since once dynamic resolution is skyped, we use that instance as a default)